tidy3d/components/autograd/functions.py

```diff
@@ -0,0 +1,91 @@
+import itertools
+
+import autograd.numpy as anp
+import numpy as np
+from numpy.typing import NDArray
```

Comment:
Yeah there are a bunch of useful types in numpy: https://numpy.org/doc/stable/reference/typing.html

`NDArray` can be further typed to specify its dtype (`NDArray[float]` for example), which might be handy sometimes, otherwise it's just `Any`